File: adapters/fastapi/src/hull_fastapi/locales.py
"""Which languages exist, and how the server picks one.

The list is **generated**. `packages/i18n` writes `locales.json` beside this
file, the same way it writes the mail bodies into `mail_templates/`, because a
tuple here beside the one in `packages/i18n/src/locales.ts` is two sources of
truth for the same fact — and the drift is silent: a locale missing from the
Python copy just quietly stops being offered.

The negotiation *algorithm* is duplicated, and that is a considered trade. The
alternative is for the browser to send its own answer at signup, which removes
the duplication and costs more than it saves: the welcome mail goes out inside
that request, so any client that forgets the field sends a Portuguese reader an
English mail, silently. The header is already in the request. Twenty lines of
RFC 4647 lookup, with the same cases asserted on both sides, is the cheaper
half. → ADR-0016
"""

from __future__ import annotations

import json
from functools import cache
from pathlib import Path

MANIFEST = Path(__file__).with_name("locales.json")


@cache
def _manifest() -> dict:
    return json.loads(MANIFEST.read_text(encoding="utf-8"))


def locales() -> tuple[str, ...]:
    return tuple(_manifest()["locales"])


def default_locale() -> str:
    return str(_manifest()["default"])


def is_locale(value: object) -> bool:
    return isinstance(value, str) and value in locales()


def resolve(value: str | None) -> str:
    """What to store for a locale somebody sent us.

    Unknown tags fall back rather than raising: this arrives from a browser and
    from `PATCH /v1/me`, and refusing a signup because the reader's language is
    not one we speak would be an odd way to say "we only have English".
    """
    return value if is_locale(value) else default_locale()


def negotiate(accept_language: str | None) -> str:
    """The best of ours for an `Accept-Language` header.

    Exact match first, then by base language — `pt`, `pt-PT` and `pt-br` all land
    on `pt-BR`, the only Portuguese we have. Serving a Portuguese reader
    Brazilian Portuguese is a small wrong; serving them English is a bigger one.
    """
    tags = _wanted(accept_language)
    available = locales()
    for tag in tags:
        for one in available:
            if one.lower() == tag:
                return one
    for tag in tags:
        base = tag.split("-")[0]
        for one in available:
            if one.lower().split("-")[0] == base:
                return one
    return default_locale()


def _wanted(header: str | None) -> list[str]:
    """Lowercased tags, best first.

    `*` is dropped: it means "anything", which is what the fallback already is.
    A malformed q-value scores 0 and is dropped with it — a header we cannot read
    is not a preference.
    """
    if not header:
        return []
    scored: list[tuple[float, int, str]] = []
    for index, part in enumerate(header.split(",")):
        tag, _, params = part.strip().partition(";")
        tag = tag.strip().lower()
        if not tag or tag == "*":
            continue
        q = 1.0
        for param in params.split(";"):
            param = param.strip()
            if param.startswith("q="):
                try:
                    q = float(param[2:])
                except ValueError:
                    q = 0.0
        if q > 0:
            # index keeps the header's own order inside equal weights.
            scored.append((-q, index, tag))
    return [tag for _, _, tag in sorted(scored)]

File: packages/i18n/src/locales.ts
/**
 * Which languages exist, and how one is chosen.
 *
 * The ladder is in ADR-0016 and has three rungs: the person's stored choice,
 * then what the browser asks for, then English. `negotiate` is the second rung
 * and nothing else — it never reads storage and never guesses from a timezone.
 */

/** English first: it is the fallback, and the order is what `check` compares against. */
export const LOCALES = ["en", "pt-BR"] as const;

export type Locale = (typeof LOCALES)[number];

export const DEFAULT_LOCALE: Locale = "en";

/** What each language calls itself. A picker that lists "Portuguese (Brazil)"
 *  in English is a picker for people who already read English. */
export const LOCALE_NAMES: Record<Locale, string> = {
  en: "English",
  "pt-BR": "Português (Brasil)",
};

export function isLocale(value: unknown): value is Locale {
  return (
    typeof value === "string" && (LOCALES as readonly string[]).includes(value)
  );
}

/**
 * Pick the best of ours for what the caller asked for.
 *
 * Takes either an `Accept-Language` header verbatim or `navigator.languages`,
 * because those are the two shapes this is ever handed and normalising them at
 * the call site would mean writing the q-value parser twice.
 */
export function negotiate(
  wanted: string | readonly string[] | null | undefined,
): Locale {
  return best(wanted, LOCALES, DEFAULT_LOCALE);
}

/**
 * The algorithm, over any list. `negotiate` is this with ours filled in.
 *
 * It takes the list rather than closing over `LOCALES` so that the exact-match
 * pass below can be *observed*. With one Portuguese in the set, deleting that
 * pass changes no answer `negotiate` can give — the near-match pass covers every
 * case — so a selftest written against `negotiate` would assert nothing while
 * appearing to assert something. This repo has shipped that mistake twice; here
 * the fix is one parameter.
 */
export function best<T extends string>(
  wanted: string | readonly string[] | null | undefined,
  available: readonly T[],
  fallback: T,
): T {
  const tags = parseWanted(wanted);
  for (const tag of tags) {
    const exact = available.find((l) => l.toLowerCase() === tag);
    if (exact) return exact;
  }
  // Then by base language: `pt`, `pt-PT` and `pt-br` all land on `pt-BR`, which
  // is the only Portuguese we have. Serving a Portuguese reader Brazilian
  // Portuguese is a small wrong; serving them English is a bigger one.
  for (const tag of tags) {
    const base = tag.split("-")[0];
    const near = available.find((l) => l.toLowerCase().split("-")[0] === base);
    if (near) return near;
  }
  return fallback;
}

/** Lowercased tags, best first. `*` is dropped: it means "anything", which is
 *  what the fallback already is. */
function parseWanted(
  wanted: string | readonly string[] | null | undefined,
): string[] {
  if (!wanted) return [];
  const raw = typeof wanted === "string" ? wanted.split(",") : [...wanted];
  return (
    raw
      .map((part) => {
        const [tag, ...params] = part.trim().split(";");
        const q = params
          .map((p) => p.trim())
          .find((p) => p.startsWith("q="))
          ?.slice(2);
        const weight = q === undefined ? 1 : Number.parseFloat(q);
        return {
          tag: tag.trim().toLowerCase(),
          q: Number.isFinite(weight) ? weight : 0,
        };
      })
      .filter((e) => e.tag && e.tag !== "*" && e.q > 0)
      // Stable within equal weights, so the header's own order still decides.
      .sort((a, b) => b.q - a.q)
      .map((e) => e.tag)
  );
}
